### ISSUE TITLE
Provide a way to hook into the rolling update

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/cluster-api#4229]
**User Story**

As a service provider, I would like to be able to hook into the rolling update of control plane
and worker nodes to be able to implement additional checks before the rolling update continues.

**Detailed Description**

Some context:  we're providing Kubernetes as a Service and are currently reimplementing our
existing solution. To provide a smooth update experience we have several checks which verify if
a Node is ready to be used by customers. In our current solution we're updating Nodes sequentially
after we updated a Node, we verify that it's completely functional and then continue the update by
deleting and creating the next Node etc. .

As far as I'm aware, there is currently no way to do this in CAPI. To me it looks like e.g. the
MachineSet controller only evaluates the Ready condition of the Node in the workload cluster when
deciding if a Machine is ready ([code](https://github.com/kubernetes-sigs/cluster-api/blob/f9db369d742747ea9791b44450b7b4be94662a09/controllers/machineset_controller.go#L643-L648)). I assume
the Machine readiness then also informs the decision when the next Machine will be updated etc..

Some examples what we're considering right now before we declare a Node ready:
* Verify that all our Node daemons are up (kube-proxy, calico, fluent-bit, metric exporter, ...)
* Verify that the CSI plugin is up and registered on the Node
* Verify that cloud controller manager reconciled loadbalancer member for services of type LoadBalancer
* Apply dynamic KubeletConfiguration and wait until it's active
* Remove our own not-ready taint so customer Pods are scheduled on the new Node until we consider the Node ready

An important point is that we always want to ensure we have minimum amount of "completely" ready Nodes.

**Anything else you would like to add:**

References:

* [Slack discussion](https://kubernetes.slack.com/archives/C8TSNPY4T/p1614064764024800)
* Potential solutions might interfere with #3153

/kind feature

**Comment by vincepri** [comment]
@sbueringer Are you interested in having this change for v1alpha4?

**Comment by sbueringer** [comment]
@vincepri Yes.

**Comment by vincepri** [comment]
There is some good aspects that we could define as part of the MachineHealthCheck (maybe?) or similar struct. For example, for generic "pod readiness", we could require that a certain number of pods show up as ready before proceeding with an update.

Then there is some customizable aspect, for things that aren't generic and would need custom code to be done, which could be done with special annotations.

**Comment by vincepri** [comment]
Nevertheless, this effort might require a small proposal to continue, and if we're aiming for v1alpha4, it should be made non-breaking and delivered potentially in a patch release.

/milestone v0.4.x

**Comment by sbueringer** [comment]
@vincepri sounds good. I'll think about it and come back with a few ideas. Just to limit the scope of the proposal a bit.

Just that I get it right. You meant specifying it as part of `MachineHealthCheck` so that those readiness criterias can be used for `MachineHealthChecks` as well as during updates? Not extending `MachineHealthChecks` *and* leveraging MachineHealthChecks during updates?

**Comment by vincepri** [comment]
> You meant specifying it as part of MachineHealthCheck so that those readiness criterias can be used for MachineHealthChecks as well as during updates? Not extending MachineHealthChecks and leveraging them through MachineHealthChecks during updates?

Yes, and if we have some data structure that it allows us to define pod-based health checks, we could use the same to define checks in MachineDeployment and KCP -- and possibly share the codebase too :)

**Comment by sbueringer** [comment]
Just a short update. I'm still interested in this. Just taking a bit of time to get more familiar with the current state of the project, to be able to better judge how this fits.

**Comment by sbueringer** [comment]
Currently not really working on it
/unassign

**Comment by killianmuldoon** [comment]
Can we mark this as done as part of #6546? @sbueringer WDYT?

**Comment by chrischdi** [comment]
👍 for ControlPlane nodes.

Regarding MachineDeployments: it would currently not be possible to do things on a per MachineDeployment base, only on "after all MachineDeployments" base. It also would not reflect state to MachineDeployments. But this points are up for discussion I think :-)

**Comment by sbueringer** [comment]
I don't think that Runtime Hooks today works for the described use case above. The idea was to do additional actions per-machine.

**Comment by sbueringer** [comment]
But I'm also fine with just closing this issue. I don't have that requirement anymore and there was no other demand for it in the last almost 1,5 years.

**Comment by killianmuldoon** [comment]
Given the feedback above let's leave it open - it seems like this might be a use case for an additional runtime hook if someone from the community is interested in it.

**Comment by faiq** [comment]
/assign @faiq

i'll try my hand at it.

**Comment by bavarianbidi** [comment]
@faiq are you still working on this? I'm very interested in a hook on a machine level

**Comment by faiq** [comment]
@bavarianbidi please by all means go ahead. i haven't had as much time to work on this as I'd like!

**Comment by fabriziopandini** [comment]
it seems there are some overlaps between what is discussed in this issue and https://github.com/kubernetes-sigs/cluster-api/pull/7647, might be better to reconcile the two efforts before moving on with implementation

**Comment by bavarianbidi** [comment]
/unassign

as i didn't work with CAPI ATM :disappointed:

I've already pitched the issue to some colleagues and shared a very rough implementation with them - so :crossed_fingers:  they will jump into this :pray:

**Comment by fabriziopandini** [comment]
The Cluster API project currently lacks enough active contributors to adequately respond to all issues and PRs.

If there is concrete interest to make this move forward and some more details to discuss about what we want to do, we can re-open

/close
